README.md
# ars_progress_dialog  
Customizable progress dialog for Flutter applications with smooth animation for background dim color and blur.

### Example
![example gif](https://raw.githubusercontent.com/arsamme/flutter-progress-dialog/main/example/doc/example.gif)

## Getting Started    
 ### Install  
Add dependency to pubspec.yaml file :   
`ars_progress_dialog: 0.0.1`    

Run this command :  
`$ flutter pub get`
  
### Import  
Import class in your project :  
`import 'package:ars_progress_dialog/ars_progress_dialog.dart';`  
  
## Showing Dialog  
### Show simple progress dialog :  

```  
ArsProgressDialog progressDialog = ArsProgressDialog(
	context,
	blur: 2,
	backgroundColor: Color(0x33000000),
	animationDuration: Duration(milliseconds: 500));
	
	progressDialog.show(); // show dialog
	progressDialog.dismiss(); //close dialog
``` 
  
### Customize loading widget :  
  
```  
 ArsProgressDialog customProgressDialog = ArsProgressDialog(
	 context,
	 blur: 2,
	 backgroundColor: Color(0x33000000),
	 loadingWidget: Container(
		 width: 150,
		 height: 150,
		 color: Colors.red,
		 child: CircularProgressIndicator(),    
    ));  
```

## Properties
| **Name**          | **Type** | **Description**                                                                                                                                                  | **Default**                    |
|-------------------|----------|------------------------------------------------------------------------------------------------------------------------------------------------------------------|--------------------------------|
| backgroundColor   | Color    | Dialog dim(background) Color                                                                                                                                     | Color (0x99000000)             |
| blur              | double   | Blur amount of dialog background                                                                                                                                 | 0                              |
| dismissable       | bool     | Setting this true lets user dismiss dialog by touching outside of it.                                                                                            | true                           |
| onDismiss         | Function | This function triggers when user dismisses dialog.                                                                                                               | -                              |
| loadingWidget     | Widget   | Dialog's widget. You can use your own widget when showing dialog.                                                                                                | _simple widget_                |
| useSafeArea       | bool     | Setting this to false makes dialog background fullscreen but when you set it true blur and background color will not apply on status bar, navigation bar and ... | false                          |
| animationDuration | Duration | This duration defines how much will take for blur and background color to appear.                                                                                | Duration (milliseconds :  300) |

CHANGELOG.md
## [0.1.2]

- Added comments

## [0.1.0]

- Add properties in README
- Removed unnecessary fields
- Changes in properties default values

## [0.0.5]

- Changed lib name, Update README

## [0.0.1]

- Initial Release

lib/ars_progress_dialog.dart
import 'dart:ui';

import 'package:flutter/material.dart';

/// Ars Progress Dialog
class ArsProgressDialog {
  /// BuildContext to show dialog in
  final BuildContext context;

  /// Main widget of dialog,
  final Widget loadingWidget;

  /// Whether dialog can dismiss by touching outside or not
  final bool dismissable;

  /// This function will trigger when user dismisses dialog
  final Function onDismiss;

  /// Amount of background blur
  final double blur;

  /// Dialog's background color
  final Color backgroundColor;

  /// Using safe area, dialog will keep padding from status bar, navigation bar and ...
  final bool useSafeArea;

  /// Duration of blur and background color animation
  final Duration animationDuration;

  /// Is dialog showing or not
  bool _isShowing = false;

  /// Dialog widget instance
  _ArsProgressDialogWidget _progressDialogWidget;

  /// Getter for _isShowing
  bool get isShowing => _isShowing;

  ArsProgressDialog(
    this.context, {
    this.backgroundColor: const Color(0x99000000),
    this.blur: 0,
    this.dismissable: true,
    this.onDismiss,
    this.loadingWidget,
    this.useSafeArea: false,
    this.animationDuration: const Duration(milliseconds: 300),
  }) {
    _initProgress();
  }

  /// Initialize dialog's instance
  void _initProgress() {
    _progressDialogWidget = _ArsProgressDialogWidget(
      blur: blur,
      dismissable: dismissable,
      backgroundColor: backgroundColor,
      onDismiss: onDismiss,
      loadingWidget: loadingWidget,
      animationDuration: animationDuration,
    );
  }

  /// Show progress dialog
  void show() async {
    if (!_isShowing) {
      _isShowing = true;
      if (_progressDialogWidget == null) _initProgress();
      await showDialog(
        useSafeArea: useSafeArea,
        context: context,
        barrierDismissible: dismissable ?? true,
        builder: (context) => _progressDialogWidget,
        barrierColor: Colors.transparent,
      );
      _isShowing = false;
    }
  }

  /// Dismiss progress dialog
  void dismiss() {
    if (_isShowing) {
      _isShowing = false;
      Navigator.pop(context);
    }
  }
}

// ignore: must_be_immutable
class _ArsProgressDialogWidget extends StatelessWidget {
  /// Main widget of dialog,
  Widget loadingWidget;

  /// This function will trigger when user dismisses dialog
  final Function onDismiss;

  /// Amount of background blur
  final double blur;

  /// Dialog's background color
  final Color backgroundColor;

  /// Whether dialog can dismiss by touching outside or not
  final bool dismissable;

  /// Duration of blur and background color animation
  final Duration animationDuration;

  _ArsProgressDialogWidget({
    this.dismissable,
    this.onDismiss,
    this.backgroundColor,
    this.loadingWidget,
    this.blur,
    this.animationDuration: const Duration(milliseconds: 300),
  }) {
    loadingWidget = loadingWidget ??
        Container(
          padding: EdgeInsets.all(10.0),
          height: 100.0,
          width: 100.0,
          alignment: Alignment.center,
          decoration: BoxDecoration(
            borderRadius: BorderRadius.circular(20),
            color: Colors.white,
          ),
          child: CircularProgressIndicator(
            strokeWidth: 2,
          ),
        );
  }

  @override
  Widget build(BuildContext context) {
    return _DialogBackground(
      blur: blur,
      dismissable: dismissable ?? true,
      onDismiss: onDismiss,
      color: backgroundColor,
      animationDuration: animationDuration,
      dialog: Padding(
        padding: MediaQuery.of(context).viewInsets +
            const EdgeInsets.symmetric(horizontal: 20.0, vertical: 24.0),
        child: Center(
          child: loadingWidget,
        ),
      ),
    );
  }
}

// ignore: must_be_immutable
class _DialogBackground extends StatelessWidget {
  /// Widget of dialog, you can use NDialog, Dialog, AlertDialog or Custom your own Dialog
  final Widget dialog;

  /// Because blur dialog cover the barrier, you have to declare here
  final bool dismissable;

  /// Action before dialog dismissed
  final Function onDismiss;

  /// Creates an background filter that applies a Gaussian blur.
  /// Default = 0
  final double blur;

  /// Background color
  final Color color;

  /// Animation Duration
  final Duration animationDuration;

  /// Color Opacity
  double _colorOpacity;

  _DialogBackground(
      {this.dialog,
      this.dismissable,
      this.blur,
      this.onDismiss,
      this.animationDuration: const Duration(milliseconds: 300),
      this.color}) {
    _colorOpacity = color.opacity;
  }

  @override
  Widget build(BuildContext context) {
    return TweenAnimationBuilder(
        tween: Tween<double>(begin: 0, end: 1),
        duration: animationDuration,
        builder: (context, val, child) {
          return Material(
            type: MaterialType.canvas,
            color: color.withOpacity(val * _colorOpacity),
            child: WillPopScope(
              onWillPop: () async {
                if (dismissable ?? true) {
                  if (onDismiss != null) onDismiss();
                  Navigator.pop(context);
                }
                return;
              },
              child: Stack(
                clipBehavior: Clip.antiAlias,
                alignment: Alignment.center,
                children: <Widget>[
                  GestureDetector(
                      onTap: dismissable ?? true
                          ? () {
                              if (onDismiss != null) {
                                onDismiss();
                              }
                              Navigator.pop(context);
                            }
                          : () {},
                      child: BackdropFilter(
                        filter: ImageFilter.blur(
                          sigmaX: val * blur,
                          sigmaY: val * blur,
                        ),
                        child: Container(
                          color: Colors.transparent,
                        ),
                      )),
                  dialog
                ],
              ),
            ),
          );
        });
  }
}


lib/custom_progress_dialog.dart
import 'dart:ui';

import 'package:flutter/material.dart';

class CustomProgressDialog {
  final BuildContext context;
  final Widget loadingWidget;
  final Function onCancel;
  final bool dismissable;
  final Function onDismiss;
  final double blur;
  final Color backgroundColor;
  bool _isShowing = false;
  ArsProgressDialogWidget _progressDialogWidget;

  bool get isShowing => _isShowing;

  CustomProgressDialog(
      this.context, {
        this.backgroundColor,
        this.blur,
        this.onCancel,
        this.dismissable,
        this.onDismiss,
        this.loadingWidget,
      }) {
    _initProgress();
  }

  void _initProgress() {
    _progressDialogWidget = ArsProgressDialogWidget(
      blur: blur,
      onCancel: onCancel,
      dismissable: dismissable,
      backgroundColor: backgroundColor,
      onDismiss: onDismiss,
      loadingWidget: loadingWidget,
    );
  }

  void show() async {
    if (!_isShowing) {
      _isShowing = true;
      if (_progressDialogWidget == null) _initProgress();
      await showDialog(
          useSafeArea: false,
          context: context,
          barrierDismissible: dismissable ?? true,
          builder: (context) => _progressDialogWidget,
          barrierColor: Color(0x00ffffff));
      _isShowing = false;
    }
  }

  ///Dissmiss progress dialog
  void dismiss() {
    if (_isShowing) {
      _isShowing = false;
      Navigator.pop(context);
    }
  }
}

// ignore: must_be_immutable
class ArsProgressDialogWidget extends StatefulWidget {
  final Function onCancel;
  final Widget loadingWidget;
  final Function onDismiss;
  final double blur;
  final Color backgroundColor;
  final bool dismissable;

  ArsProgressDialogWidget({
    Key key,
    this.onCancel,
    this.dismissable,
    this.onDismiss,
    this.backgroundColor,
    this.loadingWidget,
    this.blur,
  }) : super(key: key);

  @override
  _ArsProgressDialogWidgetState createState() =>
      _ArsProgressDialogWidgetState();
}

class _ArsProgressDialogWidgetState extends State<ArsProgressDialogWidget> {
  Widget _loadingWidget;
  Color _backgroundColor;

  @override
  Widget build(BuildContext context) {
    Color backgroundColor = _backgroundColor ?? widget.backgroundColor;
    Widget loadingWidget = (this._loadingWidget ?? widget.loadingWidget) ??
        Container(
          padding: EdgeInsets.all(10.0),
          height: 100.0,
          width: 100.0,
          alignment: Alignment.center,
          decoration: BoxDecoration(
            borderRadius: BorderRadius.circular(20),
            color: Colors.white,
          ),
          child: CircularProgressIndicator(
            strokeWidth: 2,
          ),
        );

    return DialogBackground(
      blur: widget.blur ?? 0,
      dismissable: widget.dismissable ?? true,
      onDismiss: widget.onDismiss,
      color: backgroundColor,
      dialog: Padding(
        padding: MediaQuery.of(context).viewInsets +
            const EdgeInsets.symmetric(horizontal: 20.0, vertical: 24.0),
        child: Center(
          child: loadingWidget,
        ),
      ),
    );
  }
}

class DialogBackground extends StatelessWidget {
  ///Widget of dialog, you can use NDialog, Dialog, AlertDialog or Custom your own Dialog
  final Widget dialog;

  ///Because blur dialog cover the barrier, you have to declare here
  final bool dismissable;

  ///Action before dialog dismissed
  final Function onDismiss;

  /// Creates an background filter that applies a Gaussian blur.
  /// Default = 0
  final double blur;

  /// Background color
  final Color color;

  const DialogBackground(
      {Key key,
        this.dialog,
        this.dismissable,
        this.blur,
        this.onDismiss,
        this.color})
      : super(key: key);

  ///Show dialog directly
  Future show<T>(BuildContext context) => showDialog<T>(
      context: context,
      builder: (context) => this,
      barrierColor: Color(0x00ffffff),
      barrierDismissible: dismissable ?? true);

  @override
  Widget build(BuildContext context) {
    return Material(
      type: MaterialType.canvas,
      color: color ?? Colors.black.withOpacity(.6),
      child: WillPopScope(
        onWillPop: () async {
          if (dismissable ?? true) {
            if (onDismiss != null) onDismiss();
            Navigator.pop(context);
          }
          return;
        },
        child: Stack(
          clipBehavior: Clip.antiAlias,
          alignment: Alignment.center,
          children: <Widget>[
            GestureDetector(
                onTap: dismissable ?? true
                    ? () {
                  if (onDismiss != null) {
                    onDismiss();
                  }
                  Navigator.pop(context);
                }
                    : () {},
                child: TweenAnimationBuilder(
                  tween: Tween<double>(begin: 0, end: blur ?? 0),
                  duration: Duration(milliseconds: 300),
                  builder: (context, val, child) {
                    return BackdropFilter(
                      filter: ImageFilter.blur(
                        sigmaX: val,
                        sigmaY: val,
                      ),
                      child: Container(
                        color: Colors.transparent,
                      ),
                    );
                  },
                )),
            dialog
          ],
        ),
      ),
    );
  }
}

example/lib/main.dart
import 'package:ars_progress_dialog/ars_progress_dialog.dart';
import 'package:flutter/material.dart';

void main() {
  runApp(MyApp());
}

class MyApp extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Flutter Demo',
      theme: ThemeData(
        primarySwatch: Colors.blue,
      ),
      home: MyHomePage(),
    );
  }
}

class MyHomePage extends StatefulWidget {
  @override
  _MyHomePageState createState() => _MyHomePageState();
}

class _MyHomePageState extends State<MyHomePage> {
  ArsProgressDialog _progressDialog;

  @override
  Widget build(BuildContext context) {
    _progressDialog = ArsProgressDialog(context,
        blur: 2,
        backgroundColor: Color(0x33000000),
        animationDuration: Duration(milliseconds: 500));

    ArsProgressDialog customProgressDialog = ArsProgressDialog(context,
        blur: 2,
        backgroundColor: Color(0x33000000),
        loadingWidget: Container(
          width: 150,
          height: 150,
          color: Colors.red,
          child: CircularProgressIndicator(),
        ));

    return Scaffold(
        body: Center(
      child: Column(
        mainAxisAlignment: MainAxisAlignment.center,
        children: <Widget>[
          Container(
            height: 200,
            child: Text(
              'Example of progress dialog',
              style: TextStyle(fontWeight: FontWeight.bold, fontSize: 22),
            ),
          ),
          MaterialButton(
            onPressed: () {
              _progressDialog.show();
            },
            child: Text("show dialog"),
          ),
        ],
      ),
    ));
  }
}
